windows/CodePush.cs

```diff
@@ -0,0 +1,668 @@
+        {
+            if (currentInstance == null)
+            {
+                throw new CodePushNotInitializedException("A CodePush instance has not been created yet. Have you added it to your app's list of ReactPackages?");
+            }
+
+            return currentInstance.GetJavaScriptBundleFileAsync(assetsBundleFileName).Result;
+        }
+
+        public async Task<string> GetJavaScriptBundleFileAsync(string assetsBundleFileName)
+        {
+            this.assetsBundleFileName = assetsBundleFileName;
+            string binaryJsBundleUrl = ASSETS_BUNDLE_PREFIX + assetsBundleFileName;
+            long binaryResourcesModifiedTime = await GetBinaryResourcesModifiedTime();
+            StorageFile packageFile = await codePushPackage.GetCurrentPackageBundle(this.assetsBundleFileName);
+            if (packageFile == null)
+            {
+                // There has not been any downloaded updates.
+                CodePushUtils.logBundleUrl(binaryJsBundleUrl);
+                isRunningBinaryVersion = true;
+                return binaryJsBundleUrl;
+            }
+
+            JObject packageMetadata = await codePushPackage.GetCurrentPackage();
+            long? binaryModifiedDateDuringPackageInstall = null;
+            string binaryModifiedDateDuringPackageInstallString = (string)packageMetadata[BINARY_MODIFIED_TIME_KEY];
+            if (binaryModifiedDateDuringPackageInstallString != null)
+            {
+                binaryModifiedDateDuringPackageInstall = long.Parse(binaryModifiedDateDuringPackageInstallString);
+            }
+
+            string packageAppVersion = (string)packageMetadata["appVersion"];
+
+            // TODO: test configuration
+            if (binaryModifiedDateDuringPackageInstall != null &&
+                    binaryModifiedDateDuringPackageInstall == binaryResourcesModifiedTime &&
+                    (IsUsingTestConfiguration() || appVersion.Equals(packageAppVersion)))
+            {
+                CodePushUtils.logBundleUrl(packageFile.Path);
+                isRunningBinaryVersion = false;
+                return FILE_BUNDLE_PREFIX + packageFile.Path.Replace(ApplicationData.Current.LocalFolder.Path, "").Replace("\\", "/");
+            }
+            else
+            {
+                // The binary version is newer.
+                didUpdate = false;
+                if (!mainPage.UseDeveloperSupport || !appVersion.Equals(packageAppVersion))
+                {
+                    await ClearUpdates();
+                }
+
+                CodePushUtils.logBundleUrl(binaryJsBundleUrl);
+                isRunningBinaryVersion = true;
+                return binaryJsBundleUrl;
+            }
+        }
+
+        private ApplicationDataContainer GetCodePushSettings()
+        {
+            return ApplicationData.Current.LocalSettings.CreateContainer(CODE_PUSH_PREFERENCES, ApplicationDataCreateDisposition.Always);
```

Comment:
is it necessary to keep calling create container every time we need to read or write to the settings? is it possible we could cache this and save time?